### PR TITLE
Added Service Request

### PR DESCRIPTION
**Pull request by AliabbasMerchant**
@YourTechBud There is some problem with the proto generation.
Hence, I cannot test the thing. I will try to fix it.

However, by that time, can you please check the code?
I have written the whole Service thing. Please have a look.

### CODE REVIEW

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        AtomicReference<StreamObserver<FunctionsPayload>> requestObserverRef = new AtomicReference<>();
+        CountDownLatch finishedLatch = new CountDownLatch(1);
+        StreamObserver<FunctionsPayload> streamObserver = config.stub.service(new StreamObserver<FunctionsPayload>() {
+            @Override
+            public void onNext(FunctionsPayload payload) {
+                Gson gson = new Gson();
+                if (payload.getType().equals(Constants.TYPE_SERVICE_REGISTER)) {
+                    if (payload.getId().equals(uuid)) {
+                        JsonObject params = new JsonParser().parse(payload.getParams().toStringUtf8()).getAsJsonObject();
+                        if (params.get("ack").getAsBoolean()) {
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: Throw exception if ack is false or registeration fails

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                        }
+                    }
+                } else if (payload.getType().equals(Constants.TYPE_SERVICE_REQUEST)) {
+                    FunctionsPayload.Builder toSend = FunctionsPayload.newBuilder()
+                            .setId(payload.getId())
+                            .setType(Constants.TYPE_SERVICE_REQUEST)
+                            .setService(serviceName);
+                    if (METHODS.containsKey(payload.getFunction())) {
+                        try {
+                            String answer = METHODS.get(payload.getFunction()).run(payload.getParams().toString(), payload.getAuth().toString());
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: Don't give params or auth as string. Use the class response here so the user can do getValue to convert the message to the type they want. Maybe we should rename Response to Message

Comment (YourTechBud) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: Also, the response should be of type object. You'll have to stringify it.

File: src/main/java/com/spaceuptech/space_api/utils/ServiceFunction.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.spaceuptech.space_api.utils;
+
+@FunctionalInterface
+public interface ServiceFunction {
+    String run(String params, String auth);
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/ServiceFunction.java]: None of the parameters or return type should be string. Let's use Message with that getValue thing.

Comment (YourTechBud) [reply src/main/java/com/spaceuptech/space_api/utils/ServiceFunction.java]: Can we not use `run`? Run sounds misleading. Maybe something like onNext or onInvocation?

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Service {
+    private Config config;
+    private String serviceName;
+    private final String uuid;
+    private Map<String, ServiceFunction> METHODS;
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: There is nothing special about methods. Normal camelCase convention should work

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: src/main/java/com/spaceuptech/space_api/utils/Service.java (hunk 1 above)
Comment (AliabbasMerchant) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: Ok

File: src/main/java/com/spaceuptech/space_api/utils/Service.java (hunk 2 above)
Comment (AliabbasMerchant) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: Ok

File: src/main/java/com/spaceuptech/space_api/utils/Service.java (hunk 4 above)
Comment (AliabbasMerchant) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: Yes. Right

**Comment by YourTechBud** [comment]
The auto reconnect logic is missing in this right?

File: src/main/java/com/spaceuptech/space_api/utils/Service.java (hunk 1 above)
Comment (YourTechBud) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: You aren't throwing an exception here

Comment (YourTechBud) [reply src/main/java/com/spaceuptech/space_api/utils/Service.java]: oops my bad

**Comment by AliabbasMerchant** [comment]
@YourTechBud  I have added the reconnect logic.  
(However, in the worst case, we may end up in a stack overflow due to the recursive nature of calls to `start`)

**Comment by AliabbasMerchant** [comment]
@YourTechBud  I have added the reconnect logic

**Comment by YourTechBud** [comment]
@AliabbasMerchant the stack overflow bug still exisits

**Comment by AliabbasMerchant** [comment]
Nope

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    public void registerFunc(String functionName, ServiceFunction serviceFunction) {
+        methods.put(functionName, serviceFunction);
+    }
+
+    public void start() throws Throwable {
+        while(true) {
+            finishedLatch = new CountDownLatch(1);
+            StreamObserver<FunctionsPayload> sendStream = config.stub.service(receiveStream);
+            sendStreamRef.set(sendStream);
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: Lets put the next 3 lines (106,107,108) in a method called register service

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            finishedLatch.await();
+            sendStream.onCompleted();
+            if (throwables[0] != null) {
+                if (throwables[0].getMessage().equals(RETRY)) {
+                    new Runnable() {
+                        @Override
+                        public void run() {
+                            ManagedChannel channel = (ManagedChannel) config.stub.getChannel();
+                            ConnectivityState currentState = channel.getState(true);
+                            if (!currentState.toString().equals(STATE_READY)) {
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: Else, register service. Just make sure the retry logic is intact.

File: src/main/java/com/spaceuptech/space_api/utils/Service.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        this.serviceName = serviceName;
+        this.uuid = UUID.randomUUID().toString();
+    }
+
+    public void registerFunc(String functionName, ServiceFunction serviceFunction) {
+        methods.put(functionName, serviceFunction);
+    }
+
+    public void start() throws Throwable {
+        while(true) {
```
Comment (YourTechBud) [inline src/main/java/com/spaceuptech/space_api/utils/Service.java]: You are trying to start the stream indefinitely. Shouldn't you attempt to register when the state of the channel is ready?

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Requested some changes. If you think it will break the logic with no alternative let me know.

**Comment by AliabbasMerchant** [comment]
@YourTechBud  Please have a look

**Comment by YourTechBud** [comment]
Looks good, did you test it? Like 100 requests or something?

**Comment by AliabbasMerchant** [comment]
Yup. Works fine!

**Pull request merged**